### DISCUSSION TITLE
Source Generator Dev Question

### DISCUSSION THREAD

**Discussion by thomhurst** [microsoft/testfx#3963] [Q&A] [answered]
Heya, this isn't strictly related to MSTest so delete if you like, but GitHub doesn't give me any other ways to make contact.

At the moment my source generator is fairly dumb, and relies on static methods using `[ModuleInitializer]` in order to register static objects.

I noticed your MSTest one actually generates properties instead, and then inside `SourceGeneratedTestNodesBuilder`  you collect all of these test node properties.

My question is, in `SourceGeneratedTestNodesBuilder` how do you reference/know all the generated classes that you've created? Do you have to do some form of compilation lookup? I'd look at your source code but I believe it's closed source so I can't.

I'd like TUnit to do this so I can remove static classes/methods, but just don't know how to do it.

Thanks!

**Comment by nohwnd** [comment]
What we generate in the source generator is a list with 1 entrypoint, that is pointing to all the data in the rest of the files. It is split to different files so in ideal case we need to re-generate just one file, when editing tests instead of all the code.

kinda like this: 

root => AllMyTests
  nodes for testClass1 => new TestClass1Nodes().GetNodes();
  nodes for testClass2 => new TestClass2Nodes().GetNodes();

This one entrypoint (root) is then given directly to the test framework registration that is done via testApplicationBuilder. This registration is also source generated so you can change the name of the registering class (e.g. AllMyTests) and that way have more instances in process, rather than statics. I don't think this is strictly necessary though, we could also just provide different instances of the "list" populated with different data.

**Reply by nohwnd** [reply]
If you have some specific problems point me to your code generated code please :)

**Reply by thomhurst** [reply]
> that is pointing to all the data in the rest of the files. 
> 
> kinda like this:
> 
> root => AllMyTests nodes for testClass1 => new TestClass1Nodes().GetNodes(); nodes for testClass2 => new TestClass2Nodes().GetNodes();

This is the bit I'm not sure how to do. How does `SourceGeneratedTestNodesBuilder` know the names of the generated classes `TestClass1Nodes` `TestClass2Nodes` etc?

**Reply by nohwnd** [reply, marked as answer]
The source generator grabs all classes with the test class attribute, filters them to what we actually consider tests (e.g. non-private classes), and then uses this in 2 ways.

1) to generate each class file
2) as input (via .Collect()) for the root file

What connects them is a "naming convention", that both will generate entry for each class in the incoming list, and both will name it <className>Nodes.

**Reply by thomhurst** [reply]
Thanks, that's helpful! I'm currently generating a file per test and appending a guid to avoid naming conflicts, so I don't necessarily know the name.

I'll have to switch to generating a file per class I think and then I can use a naming convention like you are.

I'll give it a go, thanks!
